subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorIntegrationTest.groovy

```diff
@@ -694,6 +694,106 @@ class WorkerExecutorIntegrationTest extends AbstractWorkerExecutorIntegrationTes
+            import org.gradle.workers.WorkAction;
+            import org.gradle.workers.WorkParameters;
+            import org.gradle.workers.WorkQueue;
+
+            import javax.inject.Inject;
+
+            public class WorkerRunningPlugin implements Plugin<Project> {
+                public void apply(Project project) {
+                    ConfigurationContainer configurations = project.getConfigurations();
+                    NamedDomainObjectProvider<Configuration> bucket = configurations.register("bucket", config -> {
+                        config.setCanBeConsumed(false);
+                        config.setCanBeResolved(false);
+                        config.getDependencies().add(project.getDependencies().create(project));
+                    });
+
+                    NamedDomainObjectProvider<Configuration> resolvable = configurations.register("resolvable", config -> {
+                        config.setCanBeConsumed(false);
+                        config.setCanBeResolved(true);
+                        config.extendsFrom(bucket.get());
+                    });
+
+                    project.getTasks().register("runAction", ActionRunningTask.class, task -> {
+                        task.getClasspath().setFrom(resolvable.get());
+                    });
+                }
+
+                public static abstract class ActionRunningTask extends DefaultTask {
+                    @Classpath
+                    public abstract ConfigurableFileCollection getClasspath();
+
+                    @Inject
+                    public abstract WorkerExecutor getWorkerExecutor();
+
+                    @TaskAction
+                    public void execute() {
+                        // Process Isolation is required here since we are testing that a
+                        // long-lived daemon process doesn't keep project files open between
+                        // separate gradle executions.
+                        WorkQueue workQueue = getWorkerExecutor().processIsolation(spec -> {
+                            spec.getClasspath().from(getClasspath());
+                        });
+                        workQueue.submit(TestAction.class, x -> {});
+                    }
+                }
+
+                public static abstract class TestAction implements WorkAction<WorkParameters.None> {
+                    @Override
+                    public void execute() {
+                        // Make the ClassLoader "dirty" by trying to load something from it
+                        assert Thread.currentThread().getContextClassLoader().getResource("/TestResource.txt") != null;
+                    }
+                }
+            }
+        """
+
+        expect:
+        // Start the action, which leaves a daemon worker process running after execution is complete.
+        succeeds("runAction")
+        // Now, while the worker is still running, clean the project build files.
+        succeeds("clean")
```

Comment:
How do we know for certain a daemon worker process is still running here?  Could this have to potential to succeed flakily and hide evidence of this feature regressing in the future?

Reply:
There are other tests that check that the workers shutdown/don't shutdown at the right point. It's OK to me if the behavior is correct, even if it's for the wrong reason.